### ISSUE TITLE
Import all from nested modules fails

### ISSUE THREAD

**Issue by iUnknwn** [microsoft/pylance-release#402]
## Environment data

-   Language Server version: v2020.9.5
-   OS and version: Win10 x64, build 1909
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.7.3, Miniconda 4.8.2.

## Expected behaviour

Pylance correctly reads modules from a `from X.Y import *` statement.

## Actual behaviour

Pylance incorrectly says the modules are not loaded. Intellisense will recommend an auto-import statement when trying to use the imported classes. 

## Logs

```
User belongs to experiment group 'ShowPlayIcon - start'
User belongs to experiment group 'ShowExtensionSurveyPrompt - enabled'
User belongs to experiment group 'DebugAdapterFactory - experiment'
User belongs to experiment group 'PtvsdWheels37 - experiment'
User belongs to experiment group 'UseTerminalToGetActivatedEnvVars - control'
User belongs to experiment group 'LocalZMQKernel - experiment'
User belongs to experiment group 'CollectLSRequestTiming - experiment'
User belongs to experiment group 'CollectNodeLSRequestTiming - experiment'
User belongs to experiment group 'EnableIPyWidgets - experiment'
User belongs to experiment group 'RunByLine - control'
User belongs to experiment group 'CustomEditorSupport - control'
User belongs to experiment group 'pythonaa'
User belongs to experiment group 'pythonTipPromptWording'
> conda --version
> pyenv root
> python3.7 c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python3.6 c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python3 c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python2 c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> py -3.7 c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> py -3.6 c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> py -3 c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> py -2 c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
Error 2020-09-22 13:42:28: Detection of Python Interpreter for Command py and args -2 failed as file  does not exist
Error 2020-09-22 13:42:29: Failed to check if file needs to be fixed [EntryNotFound (FileSystemError): Unable to read file 'c:\GitProjects\ExcelWriter\.vscode\settings.json' (EntryNotFound (FileSystemError): Error: ENOENT: no such file or directory, open 'c:\GitProjects\ExcelWriter\.vscode\settings.json')
	at _handleError (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:706:74)
	at runMicrotasks (<anonymous>)
	at processTicksAndRejections (internal/process/task_queues.js:94:5)
	at async v.readText (c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\out\client\extension.js:1:495817)
	at async p.doesFileNeedToBeFixed (c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\out\client\extension.js:56:953520)
	at async c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\out\client\extension.js:56:952647
	at async Promise.all (index 1)
	at async p.getFilesToBeFixed (c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\out\client\extension.js:56:952593)
	at async p.updateTestSettings (c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\out\client\extension.js:56:952220)] {
  code: 'FileNotFound',
  name: 'EntryNotFound (FileSystemError)'
}
Python interpreter path: .\venv\Scripts\python.exe
Starting Pylance language server.
Error 2020-09-22 13:42:29: Failed to get interpreter information for 'C:\Users\USERNAME\AppData\Local\Microsoft\WindowsApps\python3.exe' [Error: Command failed: "C:\\Users\USERNAME\AppData\Local\Microsoft\WindowsApps\python3.exe" "c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py" "c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\interpreterInfo.py"
Python was not found but can be installed from the Microsoft Store: https://go.microsoft.com/fwlink?linkID=2082640
	at ChildProcess.exithandler (child_process.js:304:12)
	at ChildProcess.emit (events.js:223:5)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:430:11)
	at Socket.emit (events.js:223:5)
	at Pipe.<anonymous> (net.js:664:12)] {
  killed: false,
  code: 9009,
  signal: null,
  cmd: '"C:\\\\Users\\USERNAME\\AppData\\Local\\Microsoft\\WindowsApps\\python3.exe" "c:\\Users\\USERNAME\\.vscode\\extensions\\ms-python.python-2020.8.109390\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\USERNAME\\.vscode\\extensions\\ms-python.python-2020.8.109390\\pythonFiles\\interpreterInfo.py"'
}
Error 2020-09-22 13:42:29: Failed to get interpreter information for 'C:\Users\USERNAME\AppData\Local\Microsoft\WindowsApps\python.exe' [Error: Command failed: "C:\\Users\USERNAME\AppData\Local\Microsoft\WindowsApps\python.exe" "c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py" "c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\interpreterInfo.py"
Python was not found but can be installed from the Microsoft Store: https://go.microsoft.com/fwlink?linkID=2082640
	at ChildProcess.exithandler (child_process.js:304:12)
	at ChildProcess.emit (events.js:223:5)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:430:11)
	at Socket.emit (events.js:223:5)
	at Pipe.<anonymous> (net.js:664:12)] {
  killed: false,
  code: 9009,
  signal: null,
  cmd: '"C:\\\\Users\\USERNAME\\AppData\\Local\\Microsoft\\WindowsApps\\python.exe" "c:\\Users\\USERNAME\\.vscode\\extensions\\ms-python.python-2020.8.109390\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\USERNAME\\.vscode\\extensions\\ms-python.python-2020.8.109390\\pythonFiles\\interpreterInfo.py"'
}
Error 2020-09-22 13:42:29: Failed to get interpreter information for 'C:\Users\USERNAME\AppData\Local\Microsoft\WindowsApps\python3.exe' [Error: Command failed: "C:\\Users\USERNAME\AppData\Local\Microsoft\WindowsApps\python3.exe" "c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py" "c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\interpreterInfo.py"
Python was not found but can be installed from the Microsoft Store: https://go.microsoft.com/fwlink?linkID=2082640
	at ChildProcess.exithandler (child_process.js:304:12)
	at ChildProcess.emit (events.js:223:5)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:430:11)
	at Socket.emit (events.js:223:5)
	at Pipe.<anonymous> (net.js:664:12)] {
  killed: false,
  code: 9009,
  signal: null,
  cmd: '"C:\\\\Users\\USERNAME\\AppData\\Local\\Microsoft\\WindowsApps\\python3.exe" "c:\\Users\\USERNAME\\.vscode\\extensions\\ms-python.python-2020.8.109390\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\USERNAME\\.vscode\\extensions\\ms-python.python-2020.8.109390\\pythonFiles\\interpreterInfo.py"'
}
Error 2020-09-22 13:42:29: Failed to get interpreter information for 'C:\Users\USERNAME\AppData\Local\Microsoft\WindowsApps\python.exe' [Error: Command failed: "C:\\Users\USERNAME\AppData\Local\Microsoft\WindowsApps\python.exe" "c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\pyvsc-run-isolated.py" "c:\Users\USERNAME\.vscode\extensions\ms-python.python-2020.8.109390\pythonFiles\interpreterInfo.py"
Python was not found but can be installed from the Microsoft Store: https://go.microsoft.com/fwlink?linkID=2082640
	at ChildProcess.exithandler (child_process.js:304:12)
	at ChildProcess.emit (events.js:223:5)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:430:11)
	at Socket.emit (events.js:223:5)
	at Pipe.<anonymous> (net.js:664:12)] {
  killed: false,
  code: 9009,
  signal: null,
  cmd: '"C:\\\\Users\\USERNAME\\AppData\\Local\\Microsoft\\WindowsApps\\python.exe" "c:\\Users\\USERNAME\\.vscode\\extensions\\ms-python.python-2020.8.109390\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\USERNAME\\.vscode\\extensions\\ms-python.python-2020.8.109390\\pythonFiles\\interpreterInfo.py"'
}
Error 2020-09-22 13:42:29: Detection of Python Interpreter for Command python3 and args  failed as file  does not exist
> conda info --json
> conda env list
> conda info --json
> conda info --json
> conda info --json

```

## Code Snippet / Additional information

You can see the pyright error flagged here:
![image](https://user-images.githubusercontent.com/1129122/93935573-ac79fd00-fcd9-11ea-920b-ce1e71731c66.png)

If I mouse over the error, pyright is flagging the item as an undefinied variable:
![image](https://user-images.githubusercontent.com/1129122/93935612-c287bd80-fcd9-11ea-812d-28b91ff1c461.png)

However, the code executes normally. Additionally, you can see the tree here:
```cmd
C:.
│   __init__.py
│
├───charters
│   │   BaseCharter.py
│   │   XYCharter.py
│   │   __init__.py
│
├───parsers
│   │   APXYParser.py
│   │   BaseParser.py
│   │   __init__.py
```

And here is the `__init__.py` for the parser sub-package:
```python
__all__ = [
    "BaseParser", 
    "APXYParser"
]
```

**Comment by erictraut** [comment]
Does `parsers.__init__.py` explicitly import `APXYParser`? It's not sufficient for `__all__` to reference it; it needs to be imported. It's possible that you're getting lucky at runtime because some other module is importing `ExcelModules.parsers.APXYParser` explicitly.

Incidentally, using wildcard imports when importing symbols from an external package is really dangerous and can easily lead to broken code when the package is updated. It's OK for throw-away, once-executed scripts. But if this is code that needs to be maintained over time, this is not a recommended practice.

**Comment by iUnknwn** [comment]
@erictraut - I'm not sure that's correct (at least as of Python 3.7) - I mocked up a quick demo project to show this: https://github.com/iUnknwn/PythonImportDemo

If you run the `test.py` in that repo, you can see it's able to access submodules in a package using a `from testpkg import *` statement, where the package itself only has an `__all__` defined (without any import statements). 

This appears consistent with the documentation: https://docs.python.org/3/tutorial/modules.html#packages

If you run pyright on the demo, you'll see the same discrepancy - executing the code normally works, but pyright flags the imported module as undefined.

**Comment by erictraut** [comment]
I stand corrected. My understanding of `__all__` was incomplete. Your sample code was very useful.

I've implemented a fix, and it will be in the next release of Pylance.

Thanks again for reporting the issue.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2020.9.6, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/master/CHANGELOG.md#202096-23-september-2020

**Comment by iUnknwn** [comment]
Thank you!
